Add: A minimal implementation of ShiftViT

ariG23498 opened the pull request:
This is a minimal implementation of [ShiftViT](https://arxiv.org/abs/2201.10801). In this paper the authors have proposed to replace the attention module of a ViT with a **NO PARAMETER** shifting operation. In this example we have tried aligning the implementation as closely as we could with the [official implementation](https://github.com/microsoft/SPACH/blob/main/models/shiftvit.py) in PyTorch. We hope this will benefit the keras community.

[Colab Notebook](https://colab.research.google.com/drive/11XA4oH8Lf7DSIbl4xMCRVd_jHOOTSvhU?usp=sharing)

CC: @ritwikraha

fchollet commented on examples/vision/shiftvit.py:
```diff
@@ -0,0 +1,871 @@
+        )
+        return config
+
+    def _calculate_loss(self, inputs, training=False):
```
Use `data` rather than `inputs` (same for `test_step`)

fchollet commented on examples/vision/shiftvit.py:
```diff
@@ -0,0 +1,871 @@
+        # steps. If not, then throw a value error.
+        if self.total_steps < self.warmup_steps:
+            raise ValueError(
+                "Total number of steps must be larger or equal to warmup steps."
```
Printing both values will make this a more helpful error message.

fchollet reviewed (commented):
Thanks for the PR! 👍 

Please format the code using `black`.

ariG23498 commented:
Thank you for the review @fchollet 

We have made the necessary changes!

fchollet commented on examples/vision/shiftvit.py:
````diff
@@ -0,0 +1,866 @@
+```
+
+In TensorFlow it would be infeasible for us to assign shifted channels to a tensor in the
+middle of a training process. This is why we have resorted to the following procedure:
````
There are several things that could be done to implement just this, but concatenation works fine as well.

ariG23498 replied:
What would be the best way according to you? This would just be a great learning curve for me.

fchollet replied:
None of the alternative ways are super satisfactory to be honest. You could use `tensor_scatter_nd_add `, you could use `tf.experimental.numpy`, you could all the XLA op `dynamic_update_slice` directly...

ariG23498 replied:
Makes sense to me! Thanks.

fchollet commented on examples/vision/shiftvit.py:
```diff
@@ -0,0 +1,866 @@
+"""
+#### Stacked Shift Blocks
+
+Each stage will have a different number of stacking as suggested in the paper. This is a
```
What does this sentence mean?

ariG23498 replied:
We meant that each stage has a variable number of stages ShiftViT Block. Should I change this sentence?

fchollet replied:
Yes, please clarify

fchollet reviewed (commented):
Thanks for the updates! I pushed some copyedits.

ariG23498 commented on examples/vision/shiftvit.py:
```diff
@@ -0,0 +1,866 @@
+These are the hyperparameters that we have chosen for the experiment. Please feel free to
+play around with them.
+
+_Note_: We do not use 4 stages as the input dimension is too small to support 4 stages
```
This is a mistake on my part. We were getting dimension error with 4 stages. We were under the impression that the dimensions of the feature map after the final stage was too small for the shifitng operation to work! I have tried with 4 stages and different setting, it works. I will be removing this note.

ariG23498 commented on examples/vision/shiftvit.py:
```diff
@@ -0,0 +1,866 @@
+MLP network is a Feed Forward Network. By default, the MLP is implemented by
+`keras.layers.Dense`. However, in this implementation, the data layout is
+in format of `[N, H, W, C]`, therefore we use 1x1 convolution to implement
+fully-connected MLP layers.
```
This comment was taken directly from the [official implementaiton](https://github.com/microsoft/SPACH/blob/main/models/shiftvit.py#L26-L29).

This was done to be as closely aligned as possible.

fchollet replied:
IMO using `Dense` seems preferable (more intuitive) unless there are performance benefits to using `Conv2D`. What does the performance look like?

ariG23498 replied:
Conv2D:
Time per epoch ~ 9sec
Test 1 accuracy ~ 75%

Dense:
Time per epoch ~ 10 sec
Test 1 accuracy ~ 74%

I also think with the `Conv2D` it is more about the optimization in the GPU. I would like to keep the `Conv2D` here mentioning that this design choice is taken from the official implementaiton. What do you think?

fchollet replied:
Since there is no significant performance difference, then I recommend we go with `Dense` since it is more intuitive and readable.

ariG23498 replied:
Cool!

ariG23498 commented on examples/vision/shiftvit.py:
```diff
@@ -0,0 +1,866 @@
+"""
+Title: Shifting operation replacing attention in ViT
```
Options:

- A Vision Transformer without Attention
- Replacing Attention with a No Parameter layer in a ViT

What do you think about them?

fchollet replied:
"A Vision Transformer without Attention" sounds great. Please update the title and then you can add the generated files. Thank you!

ariG23498 commented:
@fchollet I have made changes as requested!

ariG23498 commented:
Thanks for the thoroguh review @fchollet 

We have added the generated files after some minor edits (removing spaces and adding a period to the desctiption).

_Note_: The code was more than 300 lines, so we modified the `tutobook.py` and generated the files. Will that be a problem?

fchollet reviewed (approved):
Awesome -- thank you for the great contribution! 👍

fchollet commented:
> Note: The code was more than 300 lines, so we modified the tutobook.py and generated the files. Will that be a problem?

As much as possible, this should be avoided, but it's not a blocker.
